File: 1647-minimum-deletions-to-make-character-frequencies-unique/1647-minimum-deletions-to-make-character-frequencies-unique.cpp
class Solution {
    int updateFrequency(unordered_map<int, vector<char>>& freqToChar, char ch, int startFreq)
    {       
        int ans = 0;
        while(startFreq > 0 and freqToChar.find(startFreq) != freqToChar.end())
        {
            ans++;
            startFreq--;
        }
        if(startFreq > 0)
            freqToChar[startFreq].push_back(ch);
    return ans;
        
    }
public:
    int minDeletions(string s) {
        
       unordered_map<char, int>m;
       unordered_map<int, vector<char>>freqToChar;

        for(auto ch:s)
            m[ch]++;
        
        for(auto it : m)
            freqToChar[it.second].push_back(it.first);
        int res = 0;
        for(auto  it : freqToChar)
        {
            auto values = it.second;
            auto freq = it.first;
            
            if(values.size() > 1)
            {
                for(int i = 0; i < values.size() - 1; i++)
                {
                    res+=updateFrequency(freqToChar, values[i], freq);
                }
            }
        }
    return res;
    }
};

File: 1647-minimum-deletions-to-make-character-frequencies-unique/1647-minimum-deletions-to-make-character-frequencies-unique.py
class Solution:
    def minDeletions(self, s: str) -> int:
        m1 = defaultdict(int)
        
        for char in s:
            m1[char] += 1
            
        m2 = {}
        
        ans = 0
        for key, val in m1.items():
            if val not in m2:
                m2[val] = key
            else:
                while val in m2 and val > 0:
                    ans += 1
                    val -= 1
                m2[val] = key
            
                
        return ans
                